### PR TITLE
Switch for waypoint only reached when altitude is reached

### PR DESCRIPTION
**Pull request by FabianSchwartau**
Here is a bit of code that adds a variable to cli called nav_include_waypoint_alt, which, when enabled, forces the craft to reach the altitude of a waypoint before targeting the next waypoint. This is only tested on my X-Quad yet! I hope that this patch will currently have no effect on fixed wing crafts. I don't own a FW, so it would be nice if someone could test, that it does not affect it.

It should be possible to allow a FW to loiter until it reached the altitude, but a few things have to be considered here. Once it started loitering, it will never get much closer than the loiter radius. If it missed the WP by just 1m in altitude (w.r.t. WP radius), it should not start loitering to come out with a much greater offset. So maybe the loiter radius should be used for the decision if a WP was reached altitude wise. I am open to comments, but as I cannot test FW stuff, someone else has to implement that.

### CODE REVIEW

**Comment by hali9** [comment]
This is good idea but, have one fault. It intervenes in important mechanism - waypoint missing. I think it is possible to require altitude, keeping waypoint missing mechanism.
This also should work on airplane similarly to waypoint wait: after reaching waypoint loiter until alt will reached.

**Comment by FabianSchwartau** [comment]
Why is that important for mutirotors?
Yes, currently waypoint missing is basically disabled, when this feature is enabled - on multirotors. What I can implement is that if the waypoint is missed by some amount of distance (twice the waypoint radius for example) and also the old angle value. But I don't see why that would be required. If you enable this feature, you want to hit the waypoint exactly, no matter what. Even if the drone has to fly back a bit, it does not matter. The next waypoint will only start once you reached the current one.

**Comment by hali9** [comment]
Required altitude in wp It may not cause wp requirements at all costs. Sometimes when wind is high the copter was omit wp and this is not caused to return and fly few meters to hit wp.

**Comment by FabianSchwartau** [comment]
What do you suggest?

**Comment by hali9** [comment]
I think that we should not use 3D distance.
If the multirotor doesn't reach the point, it's hard for nothing to happen.
If the multirotor reaches a point, it should stay at that point and climb to the set height (plane should loiter around).

**Comment by FabianSchwartau** [comment]
> I think that we should not use 3D distance.

For what purpose? If you don't want to use it, just turn it off - which is also the default setting. But I need it for my application and maybe someone else can also use it.

> If the multirotor doesn't reach the point, it's hard for nothing to happen.

I don't understand that sentence. 

> If the multirotor reaches a point, it should stay at that point and climb to the set height (plane should loiter around).

That is exactly what the patch is doing - at least for multi rotor.

File: src/main/navigation/navigation.c [hunk 1]
```diff
@@ -1396,13 +1397,20 @@ static navigationFSMEvent_t navOnEnteringState_NAV_STATE_WAYPOINT_IN_PROGRESS(na
                     return NAV_FSM_EVENT_SUCCESS;   // will switch to NAV_STATE_WAYPOINT_REACHED
                 }
```
Comment (hali9) [inline src/main/navigation/navigation.c:1398]: Maybe only here just:
```
if (isWaypointReached(&posControl.activeWaypoint, false)) {
   if (navConfig()->general.flags.include_waypoint_alt)
    && fabsf(navGetCurrentActualPositionAndVelocity()->pos.z - waypoint->pos.z) > navConfig()->general.waypoint_radius) { //or new waypoint_height_range
      setDesiredPosition(&posControl.activeWaypoint.pos, 0, NAV_POS_UPDATE_XY | NAV_POS_UPDATE_Z);
      NAV_Status.state = MW_NAV_STATE_HOLD_TIMED; // I don't know can be do that
     return NAV_FSM_EVENT_NONE; // reproces state
   } else return NAV_FSM_EVENT_SUCCESS;   // will switch to NAV_STATE_WAYPOINT_REACHED
} else if (isWaypointMissed(&posControl.activeWaypoint)) {
   if (navConfig()->general.flags.nav_wp_allow_missing)
      return NAV_FSM_EVENT_SUCCESS;   // will switch to NAV_STATE_WAYPOINT_REACHED
   else return NAV_FSM_EVENT_NONE; // reproces state
}
```

**Comment by hali9** [comment]
> For what purpose? If you don't want to use it, just turn it off - which is also the default setting. But I need it for my application and maybe someone else can also use it.

But maybe I also wont use them, but I still want the "waypoint missing" functionality.

> That is exactly what the patch is doing - at least for multi rotor.

not exactly

Maybe instead of `include_waypoint_alt [tue, false]`
do `include_waypoint [none, alt, coordinate, both]`
where
none - as it was until now - default
alt - My sugestion
coordinate - as it was until now, but must reached waypoint 
both - Your sugestion

**Comment by FabianSchwartau** [comment]
> But maybe I also wont use them, but I still want the "waypoint missing" functionality.

Sorry, but I don't get it. When you turn it off, it is exactly as it is right now without the patch. So if you won't use it, you will get "waypoint missing" functionality.

> not exactly

Please explain why. I tested it. My multirotor will fly to the waypoint and wait there while climbing until it reached the altitude. 

> none - as it was until now - default
> alt - My sugestion
> coordinate - as it was until now, but must reached waypoint
> both - Your sugestion

OK, I see where you are going. First off: I don't like the names - they are confusing. I suggest to implement two different switches: One for allwing that a waypoint can be missed (e.g. nav_wp_allow_missing [ON, OFF]) and one switch which forces to reach the altitude of a waypoint (e.g. nav_wp_mode [2D, 3D] - that name is not perfect either). But that brings me back to my major problem: How to implement that a waypoint can be missed, while forcing the reach the altitude. The current check, that we just flew passed the waypoint with an angle >100° is not sufficient any more. Aspecially for FW. If the FW starts to loiter, the old condition would become true and the waypoint would assumed to be missed, although we did not reached the altitude yet. 
Suggestions? When would you concider a WP missed for a FW and a MR?

**Comment by hali9** [comment]
> How to implement that a waypoint can be missed, while forcing the reach the altitude.

Did You see My review  (must login) ?

**Comment by FabianSchwartau** [comment]
No? I guess not. What review? Where can I find it? I am logged in.

**Comment by FabianSchwartau** [comment]
I will have a close look at this this evening. Just by shortly looking at it, I don't understand it. Do you want to replace the whole patch with this code? Or do you want to add it to the patch?

**Comment by hali9** [comment]
> Do you want to replace the whole patch with this code?

In navigation.c Yes.

**Comment by FabianSchwartau** [comment]
And what does this have to do whith MW_NAV_STATE_HOLD_TIMED? hold timed is not pulled into development branch yet.

**Comment by hali9** [comment]
It is not pulled.
But this will allow the expansion of this PR to include planes.

**Comment by FabianSchwartau** [comment]
I don't see why that would include planes now. I don't want the OSD to show that I am in a WP Hold Timed, while I am actually en route to a WP (climbing). That are different states.

**Comment by hali9** [comment]
> I don't see why that would include planes now.

Because it can work generally, for planes and for mulirotors. 
In addition, the aircraft is more sensitive to wind and more often does not reach the altitude set at the waypoint.
You can define speed for multirotors and not for aircraft.
And this forces you to use several points just to reach the set height.
And so a plane could climb a circle.

> I don't want the OSD to show that I am in a WP Hold Timed, while I am actually en route to a WP (climbing). That are different states.

I was suggested a verified solution. This could be some variable that will force loiter.

**Comment by FabianSchwartau** [comment]
I had an intensive look at your code and I don't understand what you want to do with it. You want to switch into HOLD_TIMED when the WP is reached but not the altitude? How long would you like to wait there? You have to give a time for that. And as I said, I don't want the message that he is waiting at the WP while he is still climbing, that are two very different things.
And again my main question remains: When would you consider a WP missed for a FW and a MR?


> I was suggested a verified solution.

By whom? Where is it?

**Pull request closed without merging**